linera-core/src/client/validator_manager/manager.rs

```diff
@@ -0,0 +1,1260 @@
+    /// executes the operation and broadcasts the result to all waiting callers.
+    ///
+    /// This method also performs **subsumption-based deduplication**: if a larger
+    /// request that contains all the data needed by this request is already cached
+    /// or in flight, we can extract the subset result instead of making a new request.
+    ///
+    /// # Arguments
+    /// - `key`: Unique identifier for the request
+    /// - `operation`: Async closure that performs the actual request
+    ///
+    /// # Returns
+    /// The result from either the in-flight request or the newly executed operation
+    async fn deduplicated_request<T, F, Fut, N>(
+        &self,
+        key: RequestKey,
+        peer: N,
+        operation: F,
+    ) -> Result<T, NodeError>
+    where
+        T: Cacheable + Clone + Send + 'static,
+        F: FnOnce(N) -> Fut,
+        Fut: Future<Output = Result<T, NodeError>>,
+    {
+        // Check cache for exact or subsuming match
+        if let Some(result) = self.cache.get(&key).await {
+            return Ok(result);
+        }
+
+        // Check if there's an in-flight request (exact or subsuming)
+        if let Some(in_flight_match) = self.in_flight_tracker.try_subscribe(&key).await {
+            match in_flight_match {
+                InFlightMatch::Exact(Subscribed(mut receiver)) => {
+                    tracing::trace!(
+                        key = ?key,
+                        "deduplicating request (exact match) - joining existing in-flight request"
+                    );
+                    #[cfg(with_metrics)]
+                    metrics::REQUEST_CACHE_DEDUPLICATION.inc();
+                    // Wait for result from existing request
+                    match receiver.recv().await {
+                        Ok(result) => return result.as_ref().clone().map(T::from),
+                        Err(_) => {
+                            tracing::trace!(
+                                key = ?key,
+                                "in-flight request sender dropped"
+                            );
+                            // Fall through to execute a new request
+                        }
+                    }
+                }
+                InFlightMatch::Subsuming {
+                    key: subsuming_key,
+                    outcome: Subscribed(mut receiver),
+                } => {
+                    tracing::trace!(
+                    key = ?key,
+                    subsumed_by = ?subsuming_key,
+                        "deduplicating request (subsumption) - joining larger in-flight request"
+                    );
+                    #[cfg(with_metrics)]
+                    metrics::REQUEST_CACHE_DEDUPLICATION.inc();
+                    // Wait for result from the subsuming request
+                    match receiver.recv().await {
+                        Ok(result) => {
+                            match result.as_ref() {
```

Comment:
Why `as_ref()`? Can't we merge the two `match`es and do `Ok(Ok(res))` and `Ok(Err(e))`?

Reply:
it's an `Arc` and I need to extract the `RequestResult` that's inside.